cumulusci/utils/yaml/cumulusci_yml.py

```diff
@@ -1,4 +1,276 @@
+    description: str
+    attributes: Dict[str, Attribute]
+    validator: PythonClassPath = None
+
+
+class CumulusCIConfig(CCIDictModel):
+    keychain: PythonClassPath
+
+
+class Source(CCIDictModel):
+    github: URL = None
+    release: str = None
+
+
+class CumulusCLIConfig(CCIDictModel):
+    show_stacktraces: bool = None
+    plain_output: bool = None
+
+
+class CumulusCIRoot(CCIDictModel):
+    tasks: Dict[str, Task] = {}
+    flows: Dict[str, Flow] = {}
+    project: Project = {}
+    orgs: Orgs = {}
+    services: Dict[str, Service] = {}
+    cumulusci: CumulusCIConfig = None
+    plans: Dict[str, Plan] = []
+    minimum_cumulusci_version: str = None
+    sources: Dict[str, Source] = []
+    cli: CumulusCLIConfig = None
+
+
+class CumulusCIFile(CCIDictModel):
+    __root__: CumulusCIRoot
+
+
+def parse_from_yaml(source):
+    "Parse from a path, url, path-like or file-like"
+    return CumulusCIFile.parse_from_yaml(source)
+
+
+def validate_data(
+    data: Union[dict, list],
+    context: str = None,
+    on_error: callable = None,
+):
+    """Validate data which has already been loaded into a dictionary or list.
+
+    context is a string that will be used to give context to error messages.
+    on_error will be called for any validation errors with a dictionary in Pydantic error format
+
+    https://pydantic-docs.helpmanual.io/usage/models/#error-handling
+    """
+    return CumulusCIFile.validate_data(data, context=context, on_error=on_error)
+
+
+class ErrorDict(TypedDict):
+    "The structure of a Pydantic error dictionary. Google TypedDict if its new to you."
+    loc: Sequence[Union[str, int]]
+    msg: str
+    type: str
+
+
+def _log_yaml_error(logger, error: ErrorDict):
+    "Format and log a Pydantic-style error dictionary"
+    logger.warning("CumulusCI Parsing Error:")
+    loc = " -> ".join(repr(x) for x in error["loc"] if x != "__root__")
+    logger.warning("%s : %s", loc, error["msg"])
+    logger.error(
+        "NOTE: These errors will cause major problems in future versions of CumulusCI."
+    )
+    logger.error("Please correct them before May 1, 2021.")
+    logger.error(
+        "If you think your YAML has no error, please report the bug to the CumulusCI team."
+    )
+    logger.error("https://github.com/SFDO-Tooling/CumulusCI/issues/")
+
+
+def cci_safe_load(
+    source: DataInput, context: str = None, on_error: callable = None, logger=None
+):
+    """Load a CumulusCI.yml file and issue warnings for unknown structures."""
+
+    assert not (
+        on_error and logger
+    ), "Please specify either on_error or logger but not both"
+    on_error = on_error or (lambda error: _log_yaml_error(logger, error))
+
+    logger = logger or default_logger
+
+    with load_from_source(source) as (data_stream, filename):
+        data = load_yaml_data(data_stream, filename)
+        context = context or filename
+
+        try:
+            validate_data(data, context=context, on_error=on_error)
```

Comment:
I think we want to validate the final merged project config, not individual config files as they are loaded. Otherwise we can't show errors about things that are required but missing.

Reply:
https://github.com/SFDO-Tooling/CumulusCI/pull/1624#discussion_r606113698